File: clients/sdk/src/pda.ts
import { PublicKey } from "@solana/web3.js";
import { PROGRAM_IDS } from "./program-ids.js";
import type { Pubkeyish } from "./types.js";

export function toPubkey(value: Pubkeyish): PublicKey {
  return value instanceof PublicKey ? value : new PublicKey(value);
}

export function vaultPda(owner: Pubkeyish, agent: Pubkeyish): PublicKey {
  const [pda] = PublicKey.findProgramAddressSync(
    [Buffer.from("vault"), toPubkey(owner).toBuffer(), toPubkey(agent).toBuffer()],
    PROGRAM_IDS.creditVault,
  );
  return pda;
}

export function policyPda(vault: Pubkeyish): PublicKey {
  const [pda] = PublicKey.findProgramAddressSync(
    [Buffer.from("policy"), toPubkey(vault).toBuffer()],
    PROGRAM_IDS.creditVault,
  );
  return pda;
}

export function serviceRegistryPda(serviceAuthority: Pubkeyish): PublicKey {
  const [pda] = PublicKey.findProgramAddressSync(
    [Buffer.from("service"), toPubkey(serviceAuthority).toBuffer()],
    PROGRAM_IDS.reputation,
  );
  return pda;
}

export function agentProfilePda(agent: Pubkeyish): PublicKey {
  const [pda] = PublicKey.findProgramAddressSync(
    [Buffer.from("agent"), toPubkey(agent).toBuffer()],
    PROGRAM_IDS.reputation,
  );
  return pda;
}

/** PDA used by `record_payment` to track the agent ↔ service pair. The seeds
 *  are the *PDAs* of the agent profile and service registry, not the raw
 *  authority pubkeys — match the on-chain Anchor account constraint. */
export function agentServiceLinkPda(
  agentProfile: Pubkeyish,
  serviceRegistry: Pubkeyish,
): PublicKey {
  const [pda] = PublicKey.findProgramAddressSync(
    [
      Buffer.from("link"),
      toPubkey(agentProfile).toBuffer(),
      toPubkey(serviceRegistry).toBuffer(),
    ],
    PROGRAM_IDS.reputation,
  );
  return pda;
}

/** PendingSpend PDA — created by `request_spend`, consumed (and closed) by
 *  either `approve_spend` (CPIs into `spend`, transfers USDC) or
 *  `cancel_spend` (closes without transfer). Nonce is the vault's
 *  `pending_count` at request time, so every request gets a fresh PDA. */
export function pendingSpendPda(vault: Pubkeyish, nonce: number | bigint): PublicKey {
  const nonceBuf = Buffer.alloc(8);
  // u64 little-endian — matches Rust `nonce.to_le_bytes()`.
  nonceBuf.writeBigUInt64LE(typeof nonce === "bigint" ? nonce : BigInt(nonce));
  const [pda] = PublicKey.findProgramAddressSync(
    [Buffer.from("pending"), toPubkey(vault).toBuffer(), nonceBuf],
    PROGRAM_IDS.creditVault,
  );
  return pda;
}

/** Existence-as-signal: a `ReceiptUsed` PDA at `[b"receipt", hash]` proves
 *  the receipt has been recorded. `record_payment`'s `init` constraint (not
 *  `init_if_needed`) means a duplicate hash fails with `AccountAlreadyInUse`,
 *  which is the chain-side replay defence. */
export function receiptUsedPda(receiptHash: Uint8Array): PublicKey {
  if (receiptHash.length !== 32) {
    throw new Error(`receiptUsedPda expects a 32-byte hash, got ${receiptHash.length}`);
  }
  const [pda] = PublicKey.findProgramAddressSync(
    [Buffer.from("receipt"), Buffer.from(receiptHash)],
    PROGRAM_IDS.reputation,
  );
  return pda;
}

File: clients/sdk-python/src/agent_fuel/pda.py
from __future__ import annotations

from typing import Union

from solders.pubkey import Pubkey

from .constants import CREDIT_VAULT_PROGRAM_ID, REPUTATION_PROGRAM_ID

Pubkeyish = Union[Pubkey, str]


def to_pubkey(value: Pubkeyish) -> Pubkey:
    return value if isinstance(value, Pubkey) else Pubkey.from_string(value)


def vault_pda(owner: Pubkeyish, agent: Pubkeyish) -> Pubkey:
    pda, _ = Pubkey.find_program_address(
        [b"vault", bytes(to_pubkey(owner)), bytes(to_pubkey(agent))],
        CREDIT_VAULT_PROGRAM_ID,
    )
    return pda


def policy_pda(vault: Pubkeyish) -> Pubkey:
    pda, _ = Pubkey.find_program_address(
        [b"policy", bytes(to_pubkey(vault))],
        CREDIT_VAULT_PROGRAM_ID,
    )
    return pda


def service_registry_pda(service_authority: Pubkeyish) -> Pubkey:
    pda, _ = Pubkey.find_program_address(
        [b"service", bytes(to_pubkey(service_authority))],
        REPUTATION_PROGRAM_ID,
    )
    return pda


def agent_profile_pda(agent: Pubkeyish) -> Pubkey:
    pda, _ = Pubkey.find_program_address(
        [b"agent", bytes(to_pubkey(agent))],
        REPUTATION_PROGRAM_ID,
    )
    return pda


def agent_service_link_pda(agent_profile: Pubkeyish, service_registry: Pubkeyish) -> Pubkey:
    # Seeds are the PDAs themselves (not the raw authority pubkeys) — matches
    # the on-chain Anchor account constraint.
    pda, _ = Pubkey.find_program_address(
        [b"link", bytes(to_pubkey(agent_profile)), bytes(to_pubkey(service_registry))],
        REPUTATION_PROGRAM_ID,
    )
    return pda


def pending_spend_pda(vault: Pubkeyish, nonce: int) -> Pubkey:
    # u64 little-endian — matches Rust `nonce.to_le_bytes()`.
    nonce_bytes = nonce.to_bytes(8, "little", signed=False)
    pda, _ = Pubkey.find_program_address(
        [b"pending", bytes(to_pubkey(vault)), nonce_bytes],
        CREDIT_VAULT_PROGRAM_ID,
    )
    return pda


def receipt_used_pda(receipt_hash: bytes) -> Pubkey:
    if len(receipt_hash) != 32:
        raise ValueError(f"receipt_used_pda expects a 32-byte hash, got {len(receipt_hash)}")
    pda, _ = Pubkey.find_program_address(
        [b"receipt", receipt_hash],
        REPUTATION_PROGRAM_ID,
    )
    return pda
